pkg/querier/tenantfederation/merge_queryable.go

```diff
@@ -0,0 +1,309 @@
+// It is not safe to use the strings beyond the lifefime of the querier.
+// For the label "tenantLabelName" it will return all the tenant IDs available.
+// For the label "original_" + tenantLabelName it will return all the values
+// of the underlying queriers for tenantLabelName.
+func (m *mergeQuerier) LabelValues(name string) ([]string, storage.Warnings, error) {
+	if name == string(m.tenantLabelName) {
+		return m.tenantIDs, nil, nil
+	}
+
+	// ensure the name of a retained tenant id label gets handled under the
+	// origignal label name
+	if name == retainExistingPrefix+string(m.tenantLabelName) {
+		name = string(m.tenantLabelName)
+	}
+
+	return m.mergeDistinctStringSlice(func(q storage.Querier) ([]string, storage.Warnings, error) {
+		return q.LabelValues(name)
+	})
+}
+
+// LabelNames returns all the unique label names present in the underlying
+// queriers. It also add the tenantLabelName and if present in the original
+// results the "original_" tenantLabelName.
+func (m *mergeQuerier) LabelNames() ([]string, storage.Warnings, error) {
+	labelNames, warnings, err := m.mergeDistinctStringSlice(func(q storage.Querier) ([]string, storage.Warnings, error) {
+		return q.LabelNames()
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// check if the tenant label exists in the original result
+	var tenantLabelExists bool
+	for _, labelName := range labelNames {
+		if labelName == string(m.tenantLabelName) {
+			tenantLabelExists = true
+			break
+		}
+	}
+
+	if tenantLabelExists {
+		// the tenant label exists in the original result, so it had been retained
+		labelNames = append(labelNames, retainExistingPrefix+string(m.tenantLabelName))
+	} else {
+		// the tenant label was not existing, so it need to get added
+		labelNames = append(labelNames, string(m.tenantLabelName))
+	}
+	sort.Strings(labelNames)
+
+	return labelNames, warnings, nil
+
+}
+
+type stringSliceFunc func(storage.Querier) ([]string, storage.Warnings, error)
+
+// mergeDistinctStringSlice is aggregating results from stringSliceFunc calls
+// on a querier. It removes duplicates and sorts the result.
+//
+// TODO: Consider running stringSliceFunc calls concurrently
+func (m *mergeQuerier) mergeDistinctStringSlice(f stringSliceFunc) ([]string, storage.Warnings, error) {
+	var warnings storage.Warnings
+	resultMap := make(map[string]struct{})
+	for pos, tenantID := range m.tenantIDs {
+		result, resultWarnings, err := f(m.queriers[pos])
+		if err != nil {
+			return nil, nil, err
+		}
+		for _, e := range result {
+			resultMap[e] = struct{}{}
+		}
+		for _, w := range resultWarnings {
+			warnings = append(warnings, fmt.Errorf("error querying tenant id %s: %w", tenantID, w))
+		}
+	}
+
+	var result []string
+	for e := range resultMap {
+		result = append(result, e)
+	}
+	sort.Strings(result)
+	return result, warnings, nil
+}
+
+// Close releases the resources of the Querier.
+func (m *mergeQuerier) Close() error {
+	var errs prometheus.MultiError
```

Comment:
[nit] Use `github.com/prometheus/prometheus/tsdb/errors` `NewMulti()` instead please. I would also add a `faillint` to ensure none uses this `prometheus.MultiError`.

Reply:
Done exactly this in bd8c0cbe621939a0c79ef7834e6ade628266ef8e